src/parser.ts

```diff
@@ -0,0 +1,759 @@
+	parser((a) => ok(toOption(pa.run(a))));
+
+// ----- Data Structure Parsers ----- //
+
+/**
+ * Runs the parser `pa` over the value found at a given field in an object.
+ * Will fail if the input isn't an object or the field is missing.
+ *
+ * @param field The field containing the value to parse
+ * @param pa The parser to use on the field value
+ * @returns {Parser<A>} A parser for the value at the given field
+ * @example
+ * const json: unknown = JSON.parse('{ "foo": 42 }');
+ *
+ * const parserA = fieldParser('foo', numberParser); // Parser<number>
+ * const resultA = parse(parserA)(json); // Ok<number>, with value 42
+ *
+ * const parserB = fieldParser('bar', numberParser); // Parser<number>
+ * const resultB = parse(parserB)(json); // Err<string>, with 'missing field' err
+ */
+const fieldParser = <A>(field: string, pa: Parser<A>): Parser<A> =>
+	parser((a) => {
+		if (!isObject(a)) {
+			return err(
+				`Can't lookup field '${field}' on something that isn't an object`,
+			);
+		}
+
+		if (!(field in a)) {
+			return err(`Field ${field} doesn't exist in ${String(a)}`);
+		}
+
+		return pa.run(a[field]);
+	});
+
+/**
+ * Runs the parser `pa` over the value found at a given index in an array.
+ * Will fail if the input isn't an array or the index is empty.
+ *
+ * @param index The array index containing the value to parse
+ * @param pa The parser to use on the value at the given index
+ * @returns {Parser<A>} A parser for the value at the given index
+ * @example
+ * const json: unknown = JSON.parse('[41, 42, 43]');
+ *
+ * const parserA = indexParser(1, numberParser); // Parser<number>
+ * const resultA = parse(parserA)(json); // Ok<number>, with value 42
+ *
+ * const parserB = indexParser(7, numberParser); // Parser<number>
+ * const resultB = parse(parserB)(json); // Err<string>, with 'missing index' err
+ */
+const indexParser = <A>(index: number, pa: Parser<A>): Parser<A> =>
+	parser((a) => {
+		if (!Array.isArray(a)) {
+			return err(
+				`Can't lookup index ${index} on something that isn't an Array`,
+			);
+		}
+
+		const value: unknown = a[index];
+
+		if (value === undefined) {
+			return err(`Nothing found at index ${index}`);
+		}
+
+		return pa.run(value);
+	});
+
+/**
+ * A convenience for parsing values in nested objects. Takes a list of fields
+ * used to pinpoint a location within the nested objects.
+ *
+ * @param location Field names designating a location inside a nested object
+ * @param pa The parser to use on the value at the location
+ * @returns {Parser<A>} A parser for the value at the given location
+ * @example
+ * const json: unknown = JSON.parse('{ "foo": { "bar": 42 } }');
+ *
+ * const parser = locationParser(['foo', 'bar'], numberParser); // Parser<number>
+ * const result = parse(parser)(json); // Ok<number>, with value 42
+ */
+const locationParser = <A>(location: string[], pa: Parser<A>): Parser<A> =>
+	parser((a) => {
+		if (location.length === 1) {
```

Comment:
Sorry, it is Friday afternoon and my brain is struggling slightly to work this out myself.

In this function, it seems like you are recursively building a bunch of functions and executing them at the point when a `parse` is triggered.  Is it possible to the recursive build the bunch of functions when `locationParser` is first called, rather than every time a `parse` is clear.

Apologies if this makes no sense

Reply:
> it seems like you are recursively building a bunch of functions and executing them at the point when a `parse` is triggered

Yes I think so. In general this library builds a parser from a series of functions. They're only executed when `parse` is given said parser and a value to parse. This allows the parser to be re-used multiple times once it's been built.

> Is it possible to the recursive build the bunch of functions when `locationParser` is first called, rather than every time a `parse` is clear.

Sorry, I'm struggling to understand this sentence 😬🙃

Reply:
It should say "Is it possible to recursively build the bunch of functions when locationParser is first called, rather than every time a parse is called?"

Reply:
Very good spot! I've refactored it to do just that and added some tests.

It also occurred to me that the case of an empty list of parsers isn't an immediate error. It's a parser for a value that's not nested. See the tests for an example 🙂.

Reply:
That turned out to be a smaller change than I expected.

I don't know if this makes it more/less clear what is happening:
```
const locationParser = <A>(location: string[], pa: Parser<A>): Parser<A> => {
        if (location.length === 0) {
                return pa;
        }
        return fieldParser(
                location[0],
                location.length === 1 ? pa : locationParser(location.slice(1), pa),
        );
}
```

Reply:
Yeah tbf that does fit a more standard pattern for a recursive function.

Hmm, do you actually need that check for length 1 at all?

```ts
const locationParser = <A>(location: string[], pa: Parser<A>): Parser<A> => {
        if (location.length === 0) {
                return pa;
        }

        return fieldParser(location[0], locationParser(location.slice(1), pa));
}
```

which could also be expressed using [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) as:

```ts
const locationParser = <A>(location: string[], pa: Parser<A>): Parser<A> => {
	if (location.length === 0) {
		return pa;
	}

	const [head, ...tail] = location;

	return fieldParser(head, locationParser(tail, pa));
};
```

Is the `slice` version or the destructured version easier to read?

Reply:
Good point on removing the `=== 1`. I think the destructuring is slightly nicer, but there isn't much in it from my point of view.